File: prisma/refs/java/alien-dictionary.java
import java.util.*;

class Solution {
    public String alienOrder(String[] words) {
        Map<Character, Set<Character>> adj = new HashMap<>();
        Map<Character, Integer> indeg = new HashMap<>();
        for (String w : words)
            for (char c : w.toCharArray()) {
                adj.putIfAbsent(c, new HashSet<>());
                indeg.putIfAbsent(c, 0);
            }
        for (int i = 0; i + 1 < words.length; i++) {
            String a = words[i], b = words[i + 1];
            int minLen = Math.min(a.length(), b.length());
            if (a.length() > b.length() && a.startsWith(b)) return "";
            for (int j = 0; j < minLen; j++) {
                char x = a.charAt(j), y = b.charAt(j);
                if (x != y) {
                    if (adj.get(x).add(y)) indeg.merge(y, 1, Integer::sum);
                    break;
                }
            }
        }
        PriorityQueue<Character> q = new PriorityQueue<>();
        for (Map.Entry<Character, Integer> e : indeg.entrySet())
            if (e.getValue() == 0) q.add(e.getKey());
        StringBuilder out = new StringBuilder();
        while (!q.isEmpty()) {
            char c = q.poll();
            out.append(c);
            for (char nx : adj.get(c))
                if (indeg.merge(nx, -1, Integer::sum) == 0) q.add(nx);
        }
        return out.length() == indeg.size() ? out.toString() : "";
    }
}

File: prisma/refs/py/alien-dictionary.py
from collections import deque


def alien_order(words):
    adj = {c: set() for w in words for c in w}
    indeg = {c: 0 for c in adj}
    for a, b in zip(words, words[1:]):
        min_len = min(len(a), len(b))
        if len(a) > len(b) and a[:min_len] == b[:min_len]:
            return ""
        for x, y in zip(a, b):
            if x != y:
                if y not in adj[x]:
                    adj[x].add(y)
                    indeg[y] += 1
                break
    q = deque(sorted(c for c in indeg if indeg[c] == 0))
    out = []
    while q:
        c = q.popleft()
        out.append(c)
        for nx in adj[c]:
            indeg[nx] -= 1
            if indeg[nx] == 0:
                q.append(nx)
    return "".join(out) if len(out) == len(indeg) else ""
